pyomo/util/latex_printer.py

```diff
@@ -0,0 +1,1484 @@
+
+    paper_dimensions_used = {}
+    paper_dimensions_used['height'] = 11.0
+    paper_dimensions_used['width'] = 8.5
+    paper_dimensions_used['margin_left'] = 1.0
+    paper_dimensions_used['margin_right'] = 1.0
+    paper_dimensions_used['margin_top'] = 1.0
+    paper_dimensions_used['margin_bottom'] = 1.0
+
+    if paper_dimensions is not None:
+        for ky in [
+            'height',
+            'width',
+            'margin_left',
+            'margin_right',
+            'margin_top',
+            'margin_bottom',
+        ]:
+            if ky in paper_dimensions.keys():
+                paper_dimensions_used[ky] = paper_dimensions[ky]
+
+    if paper_dimensions_used['height'] >= 225:
+        raise ValueError('Paper height exceeds maximum dimension of 225')
+    if paper_dimensions_used['width'] >= 225:
+        raise ValueError('Paper width exceeds maximum dimension of 225')
+    if paper_dimensions_used['margin_left'] < 0.0:
+        raise ValueError('Paper margin_left must be greater than or equal to zero')
+    if paper_dimensions_used['margin_right'] < 0.0:
+        raise ValueError('Paper margin_right must be greater than or equal to zero')
+    if paper_dimensions_used['margin_top'] < 0.0:
+        raise ValueError('Paper margin_top must be greater than or equal to zero')
+    if paper_dimensions_used['margin_bottom'] < 0.0:
+        raise ValueError('Paper margin_bottom must be greater than or equal to zero')
+
+    if isinstance(pyomo_component, pyo.Objective):
+        objectives = [pyomo_component]
+        constraints = []
+        expressions = []
+        templatize_fcn = templatize_constraint
+        use_equation_environment = True
+        isSingle = True
+
+    elif isinstance(pyomo_component, pyo.Constraint):
+        objectives = []
+        constraints = [pyomo_component]
+        expressions = []
+        templatize_fcn = templatize_constraint
+        use_equation_environment = True
+        isSingle = True
+
+    elif isinstance(pyomo_component, pyo.Expression):
+        objectives = []
+        constraints = []
+        expressions = [pyomo_component]
+        templatize_fcn = templatize_expression
+        use_equation_environment = True
+        isSingle = True
+
+    elif isinstance(pyomo_component, (ExpressionBase, pyo.Var)):
+        objectives = []
+        constraints = []
+        expressions = [pyomo_component]
+        templatize_fcn = templatize_passthrough
+        use_equation_environment = True
+        isSingle = True
+
+    elif isinstance(pyomo_component, _BlockData):
+        objectives = [
+            obj
+            for obj in pyomo_component.component_data_objects(
+                pyo.Objective, descend_into=True, active=True
+            )
+        ]
+        constraints = [
+            con
+            for con in pyomo_component.component_objects(
+                pyo.Constraint, descend_into=True, active=True
+            )
+        ]
```

Comment:
(I think) you probably want to add `sort=SortComponents.deterministic` to these `component_data_objects` calls

Reply:
Done